File: SMO/SmoCompare/CS/SmoCompareUI/SqlObjectBrowser.cs
/*=====================================================================

  File:      SqlObjectBrowser.cs
  Summary:   Object browser window to allow object selection
  Date:         08-20-2004

---------------------------------------------------------------------
  This file is part of the Microsoft SQL Server Code Samples.
  Copyright (C) Microsoft Corporation.  All rights reserved.

  This source code is intended only as a supplement to Microsoft
  Development Tools and/or on-line documentation.  See these other
  materials for detailed information regarding Microsoft code samples.

  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
  PARTICULAR PURPOSE.
=======================================================================*/

#region Using directives

using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Reflection;

using Microsoft.SqlServer.Management.Common;
using Microsoft.SqlServer.Management.Smo;
using Microsoft.SqlServer.Management.Sdk.Sfc;
using Microsoft.SqlServer;
using Microsoft.SqlServer.MessageBox;

#endregion

namespace Microsoft.Samples.SqlServer
{
    partial class SqlObjectBrowser : Form
    {
        private Urn objUrn;
        private Server server;
        private bool connected;

        public bool Connected
        {
            get
            {
                return connected;
            }
        }

        public SqlObjectBrowser(ServerConnection serverConn)
        {
            InitializeComponent();

            // Connect to SQL Server
            server = new Server(serverConn);
            try
            {
                server.ConnectionContext.Connect();

                // In case connection succeeded we add the sql server node as root in object explorer (treeview)
                TreeNode tn = new TreeNode();
                tn.Text = server.Name;
                tn.Tag = server.Urn;
                this.objUrn = server.Urn;
                objectBrowserTreeView.Nodes.Add(tn);

                AddDummyNode(tn);

                connected = true;
            }
            catch (ConnectionException)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox();
                emb.Text = Properties.Resources.ConnectionFailed;
                emb.Show(this);
            }
            catch (ApplicationException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        public Urn Urn
        {
            get
            {
                return this.objUrn;
            }
        }

        private void selectButton_Click(object sender, EventArgs e)
        {
            if (objectBrowserTreeView.SelectedNode.Tag != null)
            {
                this.objUrn = (Urn)objectBrowserTreeView.SelectedNode.Tag;
            }

            this.Visible = false;
        }

        private void cancelCommandButton_Click(object sender, EventArgs e)
        {
            this.objUrn = null;
            this.Visible = false;
        }

        private void objectBrowserTreeView_AfterCollapse(object sender, TreeViewEventArgs e)
        {
            e.Node.Nodes.Clear();

            AddDummyNode(e.Node);
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")]
        private void objectBrowserTreeView_AfterExpand(object sender, TreeViewEventArgs e)
        {
            // Remove dummy node
            e.Node.Nodes.RemoveAt(0);
            SqlSmoObject smoObj = null;
            Urn urnNode = (Urn)e.Node.Tag;

            try
            {
                smoObj = server.GetSmoObject(urnNode);
                AddCollections(e, smoObj);
            }
            catch (UnsupportedVersionException)
            {
                // Right now don't do anything... but you might change the 
                // Icon so it will emphasize the version issue
            }
            catch (SmoException)
            {
                try
                {
                    AddItemInCollection(e, server);
                }
                catch (Exception except)
                {
                    ExceptionMessageBox emb = new ExceptionMessageBox(except);
                    emb.Show(this);
                }
            }
            catch (ApplicationException ex)
            {
                ExceptionMessageBox emb = new ExceptionMessageBox(ex);
                emb.Show(this);
            }
        }

        private static void AddItemInCollection(System.Windows.Forms.TreeViewEventArgs node, Server sqlServer)
        {
            Urn urnNode = (Urn)node.Node.Tag;
            SqlSmoObject smoObj = sqlServer.GetSmoObject(urnNode.Parent);
            TreeNode tn;
            NamedSmoObject namedObj;

            PropertyInfo p = smoObj.GetType().GetProperty(node.Node.Text);
            if (p != null)
            {
                ICollection iColl = p.GetValue(smoObj, null) as ICollection;
                if (iColl != null)
                {
                    IEnumerator enum1 = iColl.GetEnumerator();
                    while (enum1.MoveNext())
                    {
                        tn = new TreeNode();
                        namedObj = enum1.Current as NamedSmoObject;
                        if (namedObj != null)
                        {
                            tn.Text = namedObj.Name;
                        }
                        else
                        {
                            tn.Text = ((SqlSmoObject)enum1.Current).Urn;
                        }

                        tn.Tag = ((SqlSmoObject)enum1.Current).Urn;
                        node.Node.Nodes.Add(tn);

                        AddDummyNode(tn);
                    }
                }
                else
                {
                    Console.WriteLine(string.Format(
                        System.Globalization.CultureInfo.InvariantCulture,
                        Properties.Resources.PropertyNotICollection, p.Name));
                }
            }
            else
            {
                Console.WriteLine(string.Format(
                    System.Globalization.CultureInfo.InvariantCulture,
                    Properties.Resources.PropertyNotFound, node.Node.Text));
            }
        }

        /// <summary>
        /// Add the dummy node.
        /// </summary>
        /// <param name="tn"></param>
        /// <returns></returns>
        private static void AddDummyNode(TreeNode tn)
        {
            TreeNode tnD;

            tnD = new TreeNode();
            tnD.Tag = @"dummy";
            tn.Nodes.Add(tnD);
        }

        /// <summary>
        /// Adds children to the current node. basically all collection of a SqlSmoObject object
        /// </summary>
        /// <param name="node"></param>
        /// <param name="smoObj"></param>
        private static void AddCollections(System.Windows.Forms.TreeViewEventArgs node, SqlSmoObject smoObj)
        {
            TreeNode tn;
            Type t = smoObj.GetType();
            foreach (System.Reflection.PropertyInfo p in t.GetProperties())
            {
                if (p.Name != "SystemMessages"
                    && p.Name != "UserDefinedMessages")
                {
                    // If it is collection we create a node for it
                    if (p.PropertyType.IsSubclassOf(typeof(AbstractCollectionBase)))
                    {
                        tn = new TreeNode();
                        tn.Text = p.Name;
                        tn.Tag = new Urn(smoObj.Urn + @"/" + p.Name);
                        node.Node.Nodes.Add(tn);

                        AddDummyNode(tn);
                    }
                }
            }

            // Verify if we are at the server level and add the JobServer too 
            if (smoObj is Server)
            {
                tn = new TreeNode();
                tn.Text = Properties.Resources.JobServer;
                tn.Tag = new Urn(smoObj.Urn + @"/" + Properties.Resources.JobServer);
                node.Node.Nodes.Add(tn);

                AddDummyNode(tn);
            }

            smoObj = null;
        }

        private void SqlObjectBrowser_Load(object sender, EventArgs e)
        {
            // Display the first level of nodes
            if (objectBrowserTreeView.Nodes.Count > 0)
            {
                objectBrowserTreeView.Nodes[0].Expand();
            }
        }
    }
}

File: SMO/SmoCompare/VB/SmoCompareUI/SqlObjectBrowser.vb
'=====================================================================
'
'  File:    SqlObjectBrowser.vb
'  Summary: Object browser window to allow object selection
'  Date:    08-20-2004
'
'---------------------------------------------------------------------
'  This file is part of the Microsoft SQL Server Code Samples.
'  Copyright (C) Microsoft Corporation.  All rights reserved.
'
'  This source code is intended only as a supplement to Microsoft
'  Development Tools and/or on-line documentation.  See these other
'  materials for detailed information regarding Microsoft code samples.
'
'  THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
'  KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'  IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'  PARTICULAR PURPOSE.
'=======================================================================

#Region "Using directives"

Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.Reflection

Imports Microsoft.SqlServer.Management.Common
Imports Microsoft.SqlServer.Management.Smo
Imports Microsoft.SqlServer.Management.Sdk.Sfc
Imports Microsoft.SqlServer
Imports Microsoft.SqlServer.MessageBox


#End Region

Partial Class SqlObjectBrowser 'The Partial modifier is only required on one class definition per project.
    Inherits Form
    Private objUrn As Urn
    Private server As Server
    Private serverConnected As Boolean

    Public ReadOnly Property Connected() As Boolean
        Get
            Return Me.serverConnected
        End Get
    End Property

    Public Sub New(ByVal serverConn As ServerConnection)
        InitializeComponent()

        ' Connect to SQL Server
        server = New Server(serverConn)
        Try
            server.ConnectionContext.Connect()

            ' In case connection succeeded we add the sql server node as root in object explorer (treeview)
            Dim tn As New TreeNode()
            tn.Text = server.Name
            tn.Tag = server.Urn
            Me.objUrn = server.Urn
            objectBrowserTreeView.Nodes.Add(tn)

            AddDummyNode(tn)

            serverConnected = True
        Catch ex As ConnectionException
            Dim emb As New ExceptionMessageBox()
            emb.Text = My.Resources.ConnectionFailed
            emb.Show(Me)
        Catch ex As ApplicationException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        End Try
    End Sub

    Public ReadOnly Property Urn() As Urn
        Get
            Return Me.objUrn
        End Get
    End Property

    Private Sub selectButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles selectButton.Click
        If Not (objectBrowserTreeView.SelectedNode.Tag Is Nothing) Then
            Me.objUrn = CType(objectBrowserTreeView.SelectedNode.Tag, Urn)
        End If

        Me.Visible = False
    End Sub

    Private Sub cancelCommandButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cancelCommandButton.Click
        Me.objUrn = Nothing
        Me.Visible = False
    End Sub

    Private Sub objectBrowserTreeView_AfterCollapse(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles objectBrowserTreeView.AfterCollapse
        e.Node.Nodes.Clear()

        AddDummyNode(e.Node)
    End Sub

    <System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes")> _
    Private Sub objectBrowserTreeView_AfterExpand(ByVal sender As Object, ByVal e As TreeViewEventArgs) Handles objectBrowserTreeView.AfterExpand
        ' Remove dummy node
        e.Node.Nodes.RemoveAt(0)
        Dim smoObj As SqlSmoObject = Nothing
        Dim urnNode As Urn = CType(e.Node.Tag, Urn)

        Try
            smoObj = server.GetSmoObject(urnNode)
            AddCollections(e, smoObj)
        Catch ex As UnsupportedVersionException
            ' Right now don't do anything... but you might change the 
            ' Icon so it will emphasize the version issue
        Catch ex As SmoException
            Try
                AddItemInCollection(e, server)
            Catch except As Exception
                Dim emb As New ExceptionMessageBox(except)
                emb.Show(Me)
            End Try
        Catch ex As ApplicationException
            Dim emb As New ExceptionMessageBox(ex)
            emb.Show(Me)
        End Try
    End Sub

    Private Shared Sub AddItemInCollection(ByVal node As System.Windows.Forms.TreeViewEventArgs, ByVal sqlServer As Server)
        Dim urnNode As Urn = CType(node.Node.Tag, Urn)
        Dim smoObj As SqlSmoObject = sqlServer.GetSmoObject(urnNode.Parent)
        Dim tn As TreeNode
        Dim namedObj As NamedSmoObject

        Dim p As PropertyInfo = smoObj.GetType().GetProperty(node.Node.Text)
        If Not (p Is Nothing) Then
            Dim iColl As ICollection = TryCast(p.GetValue(smoObj, Nothing), ICollection)
            If Not (iColl Is Nothing) Then
                Dim enum1 As IEnumerator = iColl.GetEnumerator()
                While enum1.MoveNext()
                    tn = New TreeNode()
                    namedObj = TryCast(enum1.Current, NamedSmoObject)
                    If Not (namedObj Is Nothing) Then
                        tn.Text = namedObj.Name
                    Else
                        tn.Text = CType(enum1.Current, SqlSmoObject).Urn
                    End If

                    tn.Tag = CType(enum1.Current, SqlSmoObject).Urn
                    node.Node.Nodes.Add(tn)

                    AddDummyNode(tn)
                End While
            Else
                Console.WriteLine(String.Format(System.Globalization.CultureInfo.InvariantCulture, My.Resources.PropertyNotICollection, p.Name))
            End If
        Else
            Console.WriteLine(String.Format(System.Globalization.CultureInfo.InvariantCulture, My.Resources.PropertyNotFound, node.Node.Text))
        End If
    End Sub

    ''' <summary>
    ''' Add the dummy node.
    ''' </summary>
    ''' <param name="tn"></param>
    Private Shared Sub AddDummyNode(ByVal tn As TreeNode)
        Dim tnD As TreeNode

        tnD = New TreeNode()
        tnD.Tag = "dummy"
        tn.Nodes.Add(tnD)
    End Sub

    ''' <summary>
    ''' Adds children to the current node. basically all collection of a SqlSmoObject object
    ''' </summary>
    ''' <param name="node"></param>
    ''' <param name="smoObj"></param>
    Private Shared Sub AddCollections(ByVal node As System.Windows.Forms.TreeViewEventArgs, ByVal smoObj As SqlSmoObject)
        Dim tn As TreeNode
        Dim t As Type = smoObj.GetType()
        For Each p As System.Reflection.PropertyInfo In t.GetProperties()
            If (p.Name <> "SystemMessages" _
                AndAlso p.Name <> "UserDefinedMessages") Then
                ' If it is collection we create a node for it
                If p.PropertyType.IsSubclassOf(GetType(AbstractCollectionBase)) Then
                    tn = New TreeNode()
                    tn.Text = p.Name
                    tn.Tag = New Urn(smoObj.Urn.ToString() & "/" & p.Name)
                    node.Node.Nodes.Add(tn)

                    AddDummyNode(tn)
                End If
            End If
        Next

        ' Verify if we are at the server level and add the JobServer too 
        If TypeOf smoObj Is Server Then
            tn = New TreeNode()
            tn.Text = My.Resources.JobServer
            tn.Tag = New Urn(smoObj.Urn.ToString() & "/" & My.Resources.JobServer)
            node.Node.Nodes.Add(tn)

            AddDummyNode(tn)
        End If

        smoObj = Nothing
    End Sub

    Private Sub SqlObjectBrowser_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        ' Display the first level of nodes
        If objectBrowserTreeView.Nodes.Count > 0 Then
            objectBrowserTreeView.Nodes(0).Expand()
        End If
    End Sub
End Class
